### ISSUE TITLE
ESLintUtils.RuleTester should allow any arbitrary filename

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#4845]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/pull/54/commits/87570ca70cfce3a62190248de0980b21a420911e

**Expected Result**

The test that uses `filename: filenameDts` (`types.d.ts`) pass.

**Actual Result**

As noted in https://typescript-eslint.io/docs/development/custom-rules#testing-typed-rules, it's not supported:

```plaintext
Message:
  A fatal parsing error occurred: Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: tests/sandbox/types.d.ts.
The file must be included in at least one of the projects provided.
```

Interestingly, if you change the test to be `only: true`, it starts passing.

**Versions**

| package                    | version  |
| -------------------------- | -------- |
| `@typescript-eslint/utils` | `5.15.0` |

**Comment by bradzacher** [comment]
You need to have physical files on disk for it to work. The files can be empty but tgey need to exist.

We cpuld probably use something like `@typescript/vfs` #1891 to remove this restriction.

**Comment by JoshuaKGoldberg** [comment]
> physical files on disk

I... don't think that's the case anymore? In https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type we don't have a physical `file.ts` anywhere, and it still works.

**Comment by JoshuaKGoldberg** [comment]
Oh, and fun fact I'm learning now as I dig through: `RuleTester` tests create a watch program (`getProgramsForProjects` -> `createWatchProgram`)! I wonder if we could get slight test performance improvements by having them always use `singleRun`...

**Comment by JoshuaKGoldberg** [comment]
Ok I _think_ I'm seeing where the odd behavior comes from right now. The following is my understanding, and might not be right...

```ts
// ensure readFile reads the code being linted instead of the copy on disk
const oldReadFile = watchCompilerHost.readFile;
watchCompilerHost.readFile = (filePathIn, encoding): string | undefined => {
  const filePath = getCanonicalFileName(filePathIn);
  const fileContent =
    filePath === currentLintOperationState.filePath
      ? currentLintOperationState.code
      : oldReadFile(filePath, encoding);
  if (fileContent !== undefined) {
    parsedFilesSeenHash.set(filePath, createHash(fileContent));
  }
  return fileContent;
};
```

`currentLintOperationState`'s `.code` and `.filePath` get set whenever `getProgramsForProjects` is called. Right now that happens for the first file that gets tested (`parseForESLint` -> `parseAndGenerateServices` -> `getProgramAndAST` -> `createProjectProgram`).

```ts
const { ast, services } = parseAndGenerateServices(code, parserOptions);
```

...so, yes, I think I understand now why the virtual file system (#1891) would be a good thing 😄.

---

Fun fact! I separately fixed this particular issue by forcing `parseAndGenerateServices` to always create an isolated program. I changed it to assume `extra.singleRun === true` and `parseAndGenerateServicesCalls[options.filePath] > 1`. This feels like a dead end, though, as the isolated program only knows of the one source file.

(I realize this is probably all old news to existing project maintainers, just... good info for me!)

**Comment by bradzacher** [comment]
Okay so TIL that as long as you have the filename explicitly listed in the `files` config in the tsconfig, it will work without a physical file on disk.

This is because of the code you pointed to - we tell TS the file contents for the file. In doing so it bypasses TS's disk read - meaning you don't actually need the file on disk! As far as TS is concerned - it creates the program, file is missing, we then update the program with the new contents and now the file is "created" - after that it gets "updated".

If you have files on disk then you can just use `includes`. Idk if there are any pros/cons to either approach!

I think the files on disk version is probably better because you can "pre seed" contents and you can import files without having to parse them first? Idk that seems kinda niche.

---

Single run won't do what you want here.
Single run is designed for when each file. Is parsed by ESLint exactly once (I.e. In the CLI use case without fixers). Tests don't workike this though - they parse the same file over and over.

---

Isolated programs are potentially a problem because TS will type any `import`s as `any`. Some of our tests rely on importing types to test specific issue regression cases.

---

Setting up a VFS would be super powerful here because we could do away with the files on disk entirely - even the tsconfig! So we could have tests easily change tsconfig flags on the fly (right now we need to have multiple tsconfigs on disk to do this).

A VFS would also be super powerful as it'd be trivial to setup complex test cases for rules that rely on imported type info. (i.e. The test would be able to just "write" the files it needs to the VFS on the fly).

But it would probably be a decent chunk of effort to setup? Idk I haven't bothered to look into it fully yet!
